feat: add NVD API key

terriko opened the pull request:
This PR will add support for NVD API keys
* fixes #1428

This includes the following:
- adding the NVD notice as required by the NVD API docs
- adding `--nvd-api-key` to the command line
- modifying CVEDB to accept the api key and pass it through to NVD_API
- modifying NVD_API to accept and use the api key by adding it to the parameters sent with the request.
- documentation on how to get and use an api key
- accepting an $nvd_api_key environment variable as alternative to using the command-line flag.  This allows users of GitHub Actions (including us!) to use their built-in secrets options to keep the key private in the CI logs.
- enabled the nvd_api_key in our GitHub Actions setup -- this allows the long tests to pass without being rate limited

terriko commented:
Note to self: since I don't want to put a real API key directly into the tests, we'll probably just want to test what happens when a bad one is specified.  As you might expect, this causes the NVD API to return an error.

(I *am* going to have to plumb a real API key through github actions' secrets and make it work in pytest eventually, but that might not happen in this PR.)

terriko commented:
Okay, this should be working and ready for review.  I realized it doesn't really need separate tests yet because it's changing the main code path and I don't have any good way to put a real API key in there at the moment.

I *am* going to add code to handle bad API keys and tests for that, but I think that's going to happen in a separate PR.

terriko commented:
Pinging @BreadGenie, @anthonyharrison and @imsahil007 for review if you have time.

BreadGenie commented on doc/MANUAL.md:
```diff
@@ -259,6 +262,16 @@ A major benefit of using this NVD API is incremental updates which basically mea
+
+To get an API key, users should visit the [NVD API key request page](https://nvd.nist.gov/developers/request-an-api-key).
+
+Related: [NVD API Key Announcement](https://nvd.nist.gov/general/news/API-Key-Announcement)
```
Just a nitpick. "Key" is in small letters in other instances except here. Maybe we should stick with either of one.

terriko replied:
Good nitpick.  I was going to claim these were page titles and should be capitalized, but as it turns out I didn't actually directly copy the titles so there's no need.  I'll make them both lowercase.

imsahil007 commented:
I think it would be great if we can add a message as a suggestion with the [requesting NVD API key link](https://nvd.nist.gov/developers/request-an-api-key) in case the user is not setting this value. But that can be added later in a separate issue.
Other than that, this looks good to me.

terriko commented:
Looks like I should probably set up the NVD key in github actions so that the long tests pass here. I need to do some backend wrangling to set it up as a github secret, but I'll go start the request process for that now.

terriko commented:
> I think it would be great if we can add a message as a suggestion with the [requesting NVD API key link](https://nvd.nist.gov/developers/request-an-api-key) in case the user is not setting this value. But that can be added later in a separate issue. Other than that, this looks good to me.

Yeah, that's a good plan.  It would be ideal if we put it any time someone gets a 403 from NVD (because that's usually the indication that they're rate limited and really need it) but let me at least add a reminder to go under the boilerplate text from NVD in this PR and we can figure out some improved reminders in a future one.

terriko commented:
Fixed the key case issue (thanks @BreadGenie ) and added a "how to get a key" link (thanks @imsahil007 )

I expect to be getting my NVD API key set up in github actions .. probably tomorrow?  (I have to go through our ops team to set secret variables on the main repo.)  So hopefully I'll be able to resolve our failing long test when that's integrated.

terriko commented:
I've added support for an $nvd_api_key environment variable, then set that up to work as expected in GitHub actions.  (Thanks awesome and fast ops folk!)

This should fix the issue we've been having with rate limiting in GitHub Actions where the long tests have been failing.  At least, it fixed it for this PR, but as the issue has been a bit sporadic it's still possible that once is a coincidence.

terriko commented:
I was hoping to get a code review form @antoniogi but he's been a bit too busy and it's been over a week.  Since this should fix most of our CI issues at the moment I don't want to wait any longer, so I'm going to go ahead and merge.  Reviews always welcome later if anyone has anything else to say about the API code.
